compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -1613,42 +1613,75 @@ impl<'a> Parser<'a> {
-        } else if let PatKind::Ident(_, ident, _) = pat.kind {
-            if require_name
-                && (self.token == token::Comma
-                    || self.token == token::Lt
-                    || self.token == token::CloseDelim(token::Paren))
-            {
-                // `fn foo(a, b) {}`, `fn foo(a<x>, b<y>) {}` or `fn foo(usize, usize) {}`
-                if first_param {
-                    err.span_suggestion(
-                        pat.span,
-                        "if this is a `self` type, give it a parameter name",
-                        format!("self: {}", ident),
-                        Applicability::MaybeIncorrect,
-                    );
+        } else if require_name
+            && (self.token == token::Comma
+                || self.token == token::Lt
+                || self.token == token::CloseDelim(token::Paren))
+        {
+            let (ident, self_sugg, param_sugg, type_sugg) = match pat.kind {
+                PatKind::Ident(_, ident, _) => (
+                    ident,
+                    format!("self: {}", ident),
+                    format!("{}: TypeName", ident),
+                    format!("_: {}", ident),
+                ),
+                // Also catches `fn foo(&a)`.
+                PatKind::Ref(ref pat, mutab) => {
+                    match pat.clone().into_inner().kind {
+                        PatKind::Ident(_, ident, _) => {
+                            let mutab = mutab.prefix_str();
+                            (
+                                ident,
+                                format!("self: &{}{}", mutab, ident),
+                                format!("{}: &{}TypeName", ident, mutab),
+                                format!("_: &{}{}", mutab, ident),
+                            )
+                        }
+                        PatKind::Path(..) => {
+                            err.note("anonymous parameters are removed in the 2018 edition (see RFC 1685)");
+                            return None;
+                        }
+                        _ => return None,
+                    }
                 }
-                // Avoid suggesting that `fn foo(HashMap<u32>)` is fixed with a change to
-                // `fn foo(HashMap: TypeName<u32>)`.
-                if self.token != token::Lt {
-                    err.span_suggestion(
-                        pat.span,
-                        "if this is a parameter name, give it a type",
-                        format!("{}: TypeName", ident),
-                        Applicability::HasPlaceholders,
-                    );
+                // Also catches `fn foo(<Bar as T>::Baz)`
+                PatKind::Path(..) => {
+                    err.note("anonymous parameters are removed in the 2018 edition (see RFC 1685)");
+                    return None;
                 }
+                // Ignore other `PatKind`.
+                _ => return None,
+            };
```

Comment:
Using the snapshot strategy this can then be replaced with something like:

```rust
                PatKind::Ref(ref pat, mutab) if matches!(pat.clone().into_inner().kind, PatKind::Ident(_, ident, _)) => {
                    match pat.clone().into_inner().kind {
                        PatKind::Ident(_, ident, _) => {
                            let mutab = mutab.prefix_str();
                            (
                                ident,
                                format!("self: &{}{}", mutab, ident),
                                format!("{}: &{}TypeName", ident, mutab),
                                format!("_: &{}{}", mutab, ident),
                            )
                        }
                        _ => unreachable!(),
                    }
                }
                _ => {
                    let snapshot = self.clone();
                    match self.parse_ty_common(
                        AllowPlus::No,
                        AllowCVariadic::No,
                        RecoverQPath::No,
                        RecoverReturnSign::No,
                    ) {
                        Ok(ty) => {
                            err.span_suggestion_verbose(
                                pat.span,
                                "explicitly ignore the parameter name",
                                format!("_: {}", ty),
                                Applicability::MachineApplicable,
                            );
                            err.note("anonymous parameters are removed in the 2018 edition (see RFC 1685)");
                            return None;
                        }
                        Err(mut inner_err) => {
                            inner_err.cancel();
                            *self = snapshot;
                        }
                    }
                    return None,
                }
            };
```

Reply:
Doing that we can even go further and actually bubble up the `ty` and continue to the type checking _as if_ the user _had_ written `_: `, but that is not entirely necessary.

Reply:
Thanks for the pointer! Just copy-pasted your suggestion but the diagnostics don't seem to be changed (a6c75e8). Am I missing something?

Reply:
I think that `parse_ty_common` might not be the right thing to call then. Try using `parse_ty_for_param` (which should be almost the same thing, only `Yes` passed to all `parse_ty_common` arguments. Can you change the `inner_err.cancel();` to `inner_err.emit();`? I'm wondering whether it is taking that branch always and _why_.

Reply:
Suggestions are applied via 01adaef. So looks like the parser looks at `)` and cannot find a type.

Reply:
I've come up with another approach: call `to_ty()` and emit a suggestion if it successes, implemented via 2d99e68940087c7696a5d4919ba1456d6415fb9b.
@estebank Could you check it?

Reply:
Oh, that seems to work. Do we have a test for `fn foo(<A as B>::X, <C as D>::Y)`? I wonder how well this new method works with multiple arguments. Reviewing now, but I think it'll be ready to merge (and you can add a test if you have the time and r=me).

Reply:
Sure, added via 55bdf7f. The current rustc emits errors for each arg (https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=1a7575e24dd5fd79ad045613426147c2) and this PR's suggestion is also applied to both.